providers/azure.go

```diff
@@ -205,33 +237,66 @@ func (p *AzureProvider) prepareRedeem(redirectURL, code, codeVerifier string) (u
+
+	// In azure oauth v2 there is no resource param so add it only if V1 endpoint
+	// https://docs.microsoft.com/en-us/azure/active-directory/azuread-dev/azure-ad-endpoint-comparison#scopes-not-resources
+	if p.ProtectedResource != nil && p.ProtectedResource.String() != "" && !p.isV2Endpoint {
 		params.Add("resource", p.ProtectedResource.String())
 	}
+
 	return params, nil
 }
 
-// verifyTokenAndExtractEmail tries to extract email claim from either id_token or access token
+// extractClaimsIntoSession tries to extract email and groups claims from either id_token or access token
 // when oidc verifier is configured
-func (p *AzureProvider) verifyTokenAndExtractEmail(ctx context.Context, rawIDToken string, accessToken string) (string, error) {
-	email := ""
-
-	if rawIDToken != "" && p.Verifier != nil {
-		_, err := p.Verifier.Verify(ctx, rawIDToken)
-		// due to issues mentioned above, id_token may not be signed by AAD
-		if err == nil {
-			s, err := p.buildSessionFromClaims(rawIDToken, accessToken)
-			if err == nil {
-				email = s.Email
-			} else {
-				logger.Printf("unable to get claims from token: %v", err)
+func (p *AzureProvider) extractClaimsIntoSession(ctx context.Context, session *sessions.SessionState) error {
+
+	var innerErr error
+	var s *sessions.SessionState
+
+	// First let's verify session token
+	err := p.verifySessionToken(ctx, session)
+	if err != nil {
+		return fmt.Errorf("unable to verify token: %v", err)
+	}
+
+	// https://github.com/oauth2-proxy/oauth2-proxy/pull/914#issuecomment-782285814
+	// https://github.com/AzureAD/azure-activedirectory-library-for-java/issues/117
+	// due to above issues, id_token may not be signed by AAD
+	// in that case, we will fallback to access token
+	s, innerErr = p.buildSessionFromClaims(session.IDToken, session.AccessToken)
+	if innerErr != nil || s.Email == "" {
+		s, innerErr = p.buildSessionFromClaims(session.AccessToken, session.AccessToken)
+	}
+
+	if innerErr != nil {
+		return fmt.Errorf("unable to get claims from token: %v", innerErr)
+	}
+
+	session.Email = s.Email
+	if s.Groups != nil {
+		session.Groups = s.Groups
+	}
+
+	return nil
+}
+
+// verifySessionToken tries to validate id_token if present or access token when oidc verifier is configured
+func (p *AzureProvider) verifySessionToken(ctx context.Context, session *sessions.SessionState) error {
+	if p.Verifier != nil {
+		if session.IDToken != "" {
+			if _, err := p.Verifier.Verify(ctx, session.IDToken); err != nil {
+				logger.Printf("unable to verify ID token, fallback to access token: %v", err)
+				if _, err = p.Verifier.Verify(ctx, session.AccessToken); err != nil {
+					return fmt.Errorf("unable to verify access token: %v", err)
+				}
 			}
-		} else {
-			logger.Printf("unable to verify token: %v", err)
+		} else if _, err := p.Verifier.Verify(ctx, session.AccessToken); err != nil {
+			return fmt.Errorf("unable to verify access token: %v", err)
+
 		}
 	}
-
-	return email, nil
+	return nil
```

Comment:
This can be refactored a bit
```suggestion
	if p.Verifier == nil  {
		// Without a verifier there's no way to verify
		return nil
	}
	
	if session.IDToken != "" {
		if _, err := p.Verifier.Verify(ctx, session.IDToken); err != nil {
			logger.Printf("unable to verify ID token, fallback to access token: %v", err)
			if _, err = p.Verifier.Verify(ctx, session.AccessToken); err != nil {
				return fmt.Errorf("unable to verify access token: %v", err)
			}
		}
	} else if _, err := p.Verifier.Verify(ctx, session.AccessToken); err != nil {
		return fmt.Errorf("unable to verify access token: %v", err)
	}

	return nil
```